apps-rendering/src/server/footballContent.ts

```diff
@@ -1,66 +1,162 @@
+import { pipe } from 'lib';
 import fetch from 'node-fetch';
 import type { Response } from 'node-fetch';
+import type { Parser } from 'parser';
+import {
+	arrayParser,
+	fieldParser,
+	map,
+	map3,
+	map6,
+	map7,
+	maybe,
+	numberParser,
+	oneOf,
+	parse,
+	stringParser,
+	succeed,
+} from 'parser';
 
 type Teams = [string, string];
 
-const getFootballSelector = (
-	date: Option<Date>,
-	teamA: Option<string>,
-	teamB: Option<string>,
-): Option<string> => {
-	if (
-		date.kind === OptionKind.Some &&
-		teamA.kind === OptionKind.Some &&
-		teamB.kind === OptionKind.Some
-	) {
-		// MAPI sorts these by string value
-		const teams =
-			teamA.value < teamB.value
-				? [teamA.value, teamB.value]
-				: [teamB.value, teamA.value];
-
-		const d = date.value;
-		const year = d.getUTCFullYear();
-		const month = padZero(d.getUTCMonth() + 1);
-		const day = padZero(d.getUTCDate());
-
-		return some(`${year}-${month}-${day}_${teams[0]}_${teams[1]}`);
-	}
+const getFootballSelector = (date: Date, [teamA, teamB]: Teams): string => {
+	// MAPI sorts these by string value
+	const teams = teamA < teamB ? [teamA, teamB] : [teamB, teamA];
 
-	return none;
-};
+	const d = date;
+	const year = d.getUTCFullYear();
+	const month = padZero(d.getUTCMonth() + 1);
+	const day = padZero(d.getUTCDate());
 
-const getFootballEndpoint = (selectorId: Option<string>): Option<string> => {
-	if (selectorId.kind === OptionKind.Some) {
-		return some(
-			`https://mobile.guardianapis.com/sport/football/matches?selector=${selectorId.value}`,
-		);
-	}
-	return none;
+	return `${year}-${month}-${day}_${teams[0]}_${teams[1]}`;
 };
 
-type FootballResponse = Record<string, FootballContent>;
+const getFootballEndpoint = (selectorId: string): string =>
+	`https://mobile.guardianapis.com/sport/football/matches?selector=${selectorId}`;
+
+const optionToUndefined = <A>(optA: Option<A>): A | undefined =>
+	withDefault<A | undefined>(undefined)(optA);
+
+const stringOrUndefinedParser: Parser<string | undefined> = pipe(
+	stringParser,
+	maybe,
+	map(optionToUndefined),
+);
+
+const makeScorer = (
+	player: string,
+	timeInMinutes: number,
+	additionalInfo: undefined | string,
+): Scorer => ({
+	player,
+	timeInMinutes,
+	additionalInfo,
+});
+
+const makeFootballTeam = (
+	id: string,
+	name: string,
+	shortCode: string,
+	crestUri: string,
+	score: number,
+	scorers: Scorer[],
+): FootballTeam => ({
+	id,
+	name,
+	shortCode,
+	crestUri,
+	score,
+	scorers,
+});
+
+const makeFootballContent = (
+	id: string,
+	status: string,
+	kickOff: string,
+	competitionDisplayName: string,
+	homeTeam: FootballTeam,
+	awayTeam: FootballTeam,
+	venue: string | undefined,
+): FootballContent => ({
+	id,
+	status,
+	kickOff,
+	competitionDisplayName,
+	homeTeam,
+	awayTeam,
+	venue,
+});
+
+const scorerParser: Parser<Scorer> = map3(makeScorer)(
+	fieldParser('player', stringParser),
+	fieldParser('timeInMinutes', numberParser),
+	oneOf([
+		fieldParser('additionalInfo', stringOrUndefinedParser),
```

Comment:
Perhaps I've misunderstood, but to me this reads as:
`{"additionalInfo": "test"}` parses as `"test"`
`{"additionalInfo": 1}` parses as `undefined`
`{}` parses as `""`

Is this what is intended?

Reply:
No, you're right here. That one slipped through the net, should be the same as the other `oneOf`s

Reply:
Updated.